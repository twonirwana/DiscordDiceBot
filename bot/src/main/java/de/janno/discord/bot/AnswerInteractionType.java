package de.janno.discord.bot;

public enum AnswerInteractionType {
    none,
    hidden,
    reroll
}
