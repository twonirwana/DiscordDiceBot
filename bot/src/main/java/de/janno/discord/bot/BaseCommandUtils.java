package de.janno.discord.bot;

import de.janno.discord.bot.persistance.Mapper;
import de.janno.discord.bot.persistance.MessageDataDTO;
import de.janno.discord.bot.persistance.PersistenceManager;
import lombok.NonNull;

import javax.annotation.Nullable;
import java.util.UUID;

public class BaseCommandUtils {

    public static MessageDataDTO createEmptyMessageData(@NonNull UUID configUUID,
                                                        @Nullable Long guildId,
                                                        long channelId,
                                                        long messageId,
                                                        String commandId,
                                                        PersistenceManager persistenceManager) {
        MessageDataDTO messageDataDTO = new MessageDataDTO(configUUID, guildId, channelId, messageId, commandId, Mapper.NO_PERSISTED_STATE, null);
        //should not be needed but sometimes there is a retry ect and then there is already a state
        persistenceManager.deleteStateForMessage(channelId, messageId);
        persistenceManager.saveMessageData(messageDataDTO);
        return messageDataDTO;
    }
}
