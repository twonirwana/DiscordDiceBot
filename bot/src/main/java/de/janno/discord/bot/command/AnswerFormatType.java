package de.janno.discord.bot.command;

public enum AnswerFormatType {
    full,
    without_expression,
    only_dice,
    compact,
    minimal
}
