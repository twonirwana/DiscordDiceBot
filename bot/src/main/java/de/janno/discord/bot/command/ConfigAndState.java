package de.janno.discord.bot.command;

import lombok.NonNull;
import lombok.Value;

import java.util.UUID;

@Value
public class ConfigAndState<C extends Config, S extends StateData> {

    @NonNull
    UUID configUUID;

    @NonNull
    C config;

    @NonNull
    State<S> state;
}
