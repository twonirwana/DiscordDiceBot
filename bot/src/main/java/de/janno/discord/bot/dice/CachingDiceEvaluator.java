package de.janno.discord.bot.dice;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import de.janno.evaluator.dice.DiceEvaluator;
import de.janno.evaluator.dice.ExpressionException;
import de.janno.evaluator.dice.Roller;
import de.janno.evaluator.dice.random.NumberSupplier;
import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.Tags;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;

import java.util.concurrent.ExecutionException;

import static io.micrometer.core.instrument.Metrics.globalRegistry;

@Slf4j
public class CachingDiceEvaluator {

    private final LoadingCache<String, RollerOrError> diceRollerCache;
    private final DiceEvaluator diceEvaluator;

    public CachingDiceEvaluator(NumberSupplier numberSupplier, int maxNumberOfDice, int cacheSize) {
        diceEvaluator = new DiceEvaluator(numberSupplier, maxNumberOfDice);
        diceRollerCache = CacheBuilder.newBuilder()
                .maximumSize(cacheSize)
                .recordStats()
                .build(new CacheLoader<>() {
                    @Override
                    public @NonNull RollerOrError load(@NonNull String expression) {
                        try {
                            log.debug("create roller for: {}", expression);
                            Roller roller = diceEvaluator.buildRollSupplier(expression);
                            roller.roll();
                            return new RollerOrError(expression, roller, true, null);
                        } catch (ExpressionException | ArithmeticException e) {
                            return new RollerOrError(expression, null, false, e.getMessage());
                        }
                    }
                });
        Gauge.builder("diceEvaluator.cache", diceRollerCache::size).tags(Tags.of("stats", "size")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().requestCount()).tags(Tags.of("stats", "requests")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().hitCount()).tags(Tags.of("stats", "hit")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().missCount()).tags(Tags.of("stats", "miss")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().totalLoadTime()).tags(Tags.of("stats", "totalLoadTime")).register(globalRegistry);
        Gauge.builder("diceEvaluator.cache", () -> diceRollerCache.stats().averageLoadPenalty()).tags(Tags.of("stats", "averageLoadTime")).register(globalRegistry);
    }

    public @NonNull RollerOrError get(@NonNull String expression) {
        try {
            return diceRollerCache.get(expression);
        } catch (ExecutionException e) {
            throw new IllegalStateException(e);
        }
    }

    @VisibleForTesting
    public long getCacheSize() {
        return diceRollerCache.size();
    }
}
