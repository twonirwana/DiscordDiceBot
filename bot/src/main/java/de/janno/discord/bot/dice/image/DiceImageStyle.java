package de.janno.discord.bot.dice.image;

import de.janno.discord.bot.dice.image.provider.*;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public enum DiceImageStyle {
    none(new NoneProvider()),
    polyhedral_3d(new PolyhedralFileImageProvider("polyhedral_3d", List.of("red_and_white"), "red_and_white")),
    polyhedral_alies_v2(new PolyhedralFileImageProvider(
            "polyhedral_alies_v2",
            List.of("blue_and_silver", "black_and_gold", "blue_and_gold", "green_and_gold", "orange_and_silver", "red_and_gold", "purple_and_silver"),
            "blue_and_silver")),
    polyhedral_knots(new PolyhedralFileImageProvider("polyhedral_knots", List.of("purple_dark", "purple_white", "blue"), "blue")),
    polyhedral_RdD(new PolyhedralFileImageProvider("polyhedral_RdD", List.of("default", "special"), "default", Map.of(
            "default", List.of(4, 6, 7, 8, 10, 12, 20, 100),
            "special", List.of(8, 12)))),
    fate(new FateBasic()),
    d6_dots(new D6Dotted()),
    polyhedral_2d(new PolyhedralSvgWithColor()),
    polyhedral_alies_v1(new PolyhedralFileImageProvider("polyhedral_alies_v1", List.of("black_and_gold"), "black_and_gold"));

    private final ImageProvider imageProvider;

    DiceImageStyle(ImageProvider imageProvider) {
        this.imageProvider = imageProvider;
    }

    public static boolean isValidStyle(String name) {
        return Arrays.stream(values()).map(Enum::name).collect(Collectors.toSet()).contains(name);
    }

    public static String combineStyleAndColorName(DiceImageStyle style, String color) {
        return "%s_%s".formatted(style.name(), color);
    }

    public ImageProvider getImageProvider() {
        return imageProvider;
    }

    public String getDefaultColor() {
        return imageProvider.getDefaultColor();
    }

    public List<String> getSupportedColors() {
        return imageProvider.getSupportedColors();
    }
}
