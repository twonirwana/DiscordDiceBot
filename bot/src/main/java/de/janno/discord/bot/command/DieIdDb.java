package de.janno.discord.bot.command;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import de.janno.evaluator.dice.DieId;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NonNull;
import lombok.ToString;

import java.io.Serializable;

@EqualsAndHashCode
@ToString
@Getter
public class DieIdDb implements Serializable {
    private final int expressionPositionStartInc;
    @NonNull
    private final String value;
    private final int reEvaluateCounter;
    private final int dieIndex;
    private final int reroll;

    @JsonCreator
    public DieIdDb(@JsonProperty("expressionPositionStartInc") int expressionPositionStartInc,
                   @JsonProperty("value") @NonNull String value,
                   @JsonProperty("reEvaluateCounter") int reEvaluateCounter,
                   @JsonProperty("dieIndex") int dieIndex,
                   @JsonProperty("reroll") int reroll
    ) {
        this.expressionPositionStartInc = expressionPositionStartInc;
        this.value = value;
        this.reEvaluateCounter = reEvaluateCounter;
        this.dieIndex = dieIndex;
        this.reroll = reroll;
    }

    @JsonIgnore
    public static DieIdDb fromDieId(DieId dieId) {
        return new DieIdDb(
                dieId.getRollId().getExpressionPosition().getStartInc(),
                dieId.getRollId().getExpressionPosition().getValue(),
                dieId.getRollId().getReevaluate(),
                dieId.getDieIndex(),
                dieId.getReroll()
        );
    }

    @JsonIgnore
    public DieId toDieId() {
        return DieId.of(expressionPositionStartInc, value, reEvaluateCounter, dieIndex, reroll);
    }
}
