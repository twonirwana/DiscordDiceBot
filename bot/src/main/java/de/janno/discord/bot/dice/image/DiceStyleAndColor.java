package de.janno.discord.bot.dice.image;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.google.common.base.Strings;
import lombok.NonNull;
import lombok.Value;
import org.jetbrains.annotations.Nullable;

import java.awt.image.BufferedImage;
import java.util.List;
import java.util.Optional;


@Value
public class DiceStyleAndColor {
    @NonNull
    DiceImageStyle diceImageStyle;
    @NonNull
    String configuredDefaultColor;

    public DiceStyleAndColor(@NonNull @JsonProperty("diceImageStyle") DiceImageStyle diceImageStyle,
                             @Nullable @JsonProperty("configuredDefaultColor") String configuredDefaultColor) {
        this.diceImageStyle = diceImageStyle;
        this.configuredDefaultColor = Optional.ofNullable(configuredDefaultColor)
                .map(c -> {
                    if (diceImageStyle.getSupportedColors().contains(c)) {
                        return c;
                    }
                    return diceImageStyle.getDefaultColor();
                })
                .orElse(diceImageStyle.getDefaultColor());
    }

    public @NonNull List<BufferedImage> getImageFor(@Nullable Integer totalDieSides,
                                                    @Nullable Integer shownDieSide,
                                                    @Nullable String rollColor) {

        return diceImageStyle.getImageProvider().getImageFor(totalDieSides, shownDieSide, Optional.ofNullable(rollColor)
                .filter(c -> !Strings.isNullOrEmpty(c))
                .orElse(configuredDefaultColor));
    }

    @Override
    public String toString() {
        return DiceImageStyle.combineStyleAndColorName(diceImageStyle, configuredDefaultColor);
    }
}
