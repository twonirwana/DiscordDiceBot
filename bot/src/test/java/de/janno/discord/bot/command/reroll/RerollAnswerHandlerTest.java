package de.janno.discord.bot.command.reroll;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import de.janno.discord.bot.AnswerInteractionType;
import de.janno.discord.bot.command.AnswerFormatType;
import de.janno.discord.bot.command.ConfigAndState;
import de.janno.discord.bot.command.DieIdDb;
import de.janno.discord.bot.dice.CachingDiceEvaluator;
import de.janno.discord.bot.dice.image.DiceImageStyle;
import de.janno.discord.bot.dice.image.DiceStyleAndColor;
import de.janno.discord.bot.persistance.*;
import de.janno.evaluator.dice.DieId;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.util.List;
import java.util.Locale;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(SnapshotExtension.class)
class RerollAnswerHandlerTest {

    RerollAnswerHandler underTest;
    private Expect expect;

    @Test
    void checkPersistence() {
        PersistenceManager persistenceManager = new PersistenceManagerImpl("jdbc:h2:mem:" + UUID.randomUUID(), null, null);
        underTest = new RerollAnswerHandler(persistenceManager, new CachingDiceEvaluator((minExcl, maxIncl) -> 0));

        UUID configUUID = UUID.randomUUID();
        long channelId = System.currentTimeMillis();
        long messageId = System.currentTimeMillis();
        DieId dieId1 = DieId.of(1, "d", 0, 0, 0);
        DieId dieId2 = DieId.of(1, "d", 0, 1, 0);
        RerollAnswerConfig config = new RerollAnswerConfig(123L, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH, AnswerInteractionType.reroll,
                "2d6", List.of(
                new DieIdTypeAndValue(DieIdDb.fromDieId(dieId1), "2", 2, 6, null),
                new DieIdTypeAndValue(DieIdDb.fromDieId(dieId2), "5", 5, 6, null)
        ), 0, "userName", true, "roll");
        Optional<MessageConfigDTO> toSave = RerollAnswerHandler.createMessageConfig(configUUID, 1L, channelId, config);
        assertThat(toSave).isPresent();

        persistenceManager.saveMessageConfig(toSave.get());
        MessageConfigDTO loaded = persistenceManager.getMessageConfig(configUUID).orElseThrow();

        assertThat(toSave.get()).isEqualTo(loaded);
        RerollAnswerStateData stateData = new RerollAnswerStateData(List.of());
        MessageDataDTO messageDataDTO = new MessageDataDTO(configUUID, 1L, channelId, messageId, "reroll_answer", "RerollAnswerStateData", Mapper.serializedObject(stateData));

        ConfigAndState<RerollAnswerConfig, RerollAnswerStateData> configAndState = underTest.getMessageDataAndUpdateWithButtonValue(loaded, messageDataDTO, dieId1.toString(), "userName");
        assertThat(configAndState.getConfig()).isEqualTo(config);
        assertThat(configAndState.getConfigUUID()).isEqualTo(configUUID);
        assertThat(configAndState.getState().getData()).isEqualTo(new RerollAnswerStateData(List.of(DieIdDb.fromDieId(dieId1))));
    }

    @Test
    void deserialization() {
        PersistenceManager persistenceManager = new PersistenceManagerImpl("jdbc:h2:mem:" + UUID.randomUUID(), null, null);
        underTest = new RerollAnswerHandler(persistenceManager, new CachingDiceEvaluator((minExcl, maxIncl) -> 0));

        UUID configUUID = UUID.randomUUID();
        MessageConfigDTO messageConfigDTO = new MessageConfigDTO(configUUID, 1L, 1660644934298L, "reroll_answer", "RerollAnswerConfig", """
                ---
                answerTargetChannelId: 123
                answerFormatType: "full"
                configLocale: "en"
                answerInteractionType: "reroll"
                expression: "2d6"
                dieIdTypeAndValues:
                - dieIdDb:
                    expressionPositionStartInc: 1
                    value: "d"
                    reEvaluateCounter: 0
                    dieIndex: 0
                    reroll: 0
                  value: "2"
                  numberSupplierValue: 2
                  diceSides: 6
                  selectedFrom: null
                - dieIdDb:
                    expressionPositionStartInc: 1
                    value: "d"
                    reEvaluateCounter: 0
                    dieIndex: 1
                    reroll: 0
                  value: "5"
                  numberSupplierValue: 5
                  diceSides: 6
                  selectedFrom: null
                rerollCount: 0
                owner: "owner"
                alwaysSumUp: true
                label: "roll"
                diceStyleAndColor:
                  diceImageStyle: "none"
                  configuredDefaultColor: "none"
                """);
        MessageDataDTO messageDataDTO = new MessageDataDTO(configUUID, 1L, 1660644934298L, 1660644934298L, "reroll_answer",
                "RerollAnswerStateData", """
                ---
                rerollDice:
                - expressionPositionStartInc: 1
                  value: "d"
                  reEvaluateCounter: 0
                  dieIndex: 0
                  reroll: 0
                """);

        ConfigAndState<RerollAnswerConfig, RerollAnswerStateData> configAndState = underTest.getMessageDataAndUpdateWithButtonValue(messageConfigDTO, messageDataDTO, "dadf", "userName");
        RerollAnswerConfig config = new RerollAnswerConfig(123L, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH, AnswerInteractionType.reroll,
                "2d6", List.of(
                new DieIdTypeAndValue(DieIdDb.fromDieId(DieId.of(1, "d", 0, 0, 0)), "2", 2, 6, null),
                new DieIdTypeAndValue(DieIdDb.fromDieId(DieId.of(1, "d", 0, 1, 0)), "5", 5, 6, null)
        ), 0, "owner", true, "roll");
        assertThat(configAndState.getConfig()).isEqualTo(config);
        assertThat(configAndState.getConfigUUID()).isEqualTo(configUUID);
        assertThat(configAndState.getState().getData()).isEqualTo(new RerollAnswerStateData(List.of(DieIdDb.fromDieId(DieId.of(1, "d", 0, 0, 0)))));
    }

    @Test
    void configSerialization() {
        UUID configUUID = UUID.fromString("00000000-0000-0000-0000-000000000000");
        RerollAnswerConfig config = new RerollAnswerConfig(123L, AnswerFormatType.full, null, new DiceStyleAndColor(DiceImageStyle.none, "none"), Locale.ENGLISH, AnswerInteractionType.reroll,
                "2d6", List.of(
                new DieIdTypeAndValue(DieIdDb.fromDieId(DieId.of(1, "d", 0, 0, 0)), "2", 2, 6, null),
                new DieIdTypeAndValue(DieIdDb.fromDieId(DieId.of(1, "d", 0, 1, 0)), "5", 5, 6, null)
        ), 0, "owner", true, "roll");
        Optional<MessageConfigDTO> toSave = RerollAnswerHandler.createMessageConfig(configUUID, 1L, 2L, config);
        assertThat(toSave).isPresent();

        expect.toMatchSnapshot(toSave.get());
    }

}