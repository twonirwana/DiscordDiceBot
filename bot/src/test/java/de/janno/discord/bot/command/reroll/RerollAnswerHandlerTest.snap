de.janno.discord.bot.command.reroll.RerollAnswerHandlerTest.configSerialization=[
  {
    "configUUID": "00000000-0000-0000-0000-000000000000",
    "guildId": 1,
    "channelId": 2,
    "commandId": "reroll_answer",
    "configClassId": "RerollAnswerConfig",
    "config": "---\nanswerTargetChannelId: 123\nanswerFormatType: \"full\"\nconfigLocale: \"en\"\nanswerInteractionType: \"reroll\"\nexpression: \"2d6\"\ndieIdTypeAndValues:\n- dieIdDb:\n    expressionPositionStartInc: 1\n    value: \"d\"\n    reEvaluateCounter: 0\n    dieIndex: 0\n    reroll: 0\n  value: \"2\"\n  numberSupplierValue: 2\n  diceSides: 6\n  selectedFrom: null\n- dieIdDb:\n    expressionPositionStartInc: 1\n    value: \"d\"\n    reEvaluateCounter: 0\n    dieIndex: 1\n    reroll: 0\n  value: \"5\"\n  numberSupplierValue: 5\n  diceSides: 6\n  selectedFrom: null\nrerollCount: 0\nowner: \"owner\"\nalwaysSumUp: true\nlabel: \"roll\"\ndiceStyleAndColor:\n  diceImageStyle: \"none\"\n  configuredDefaultColor: \"none\"\ncallStarterConfigAfterFinish: null\nname: null\n",
    "creationUserId": 0
  }
]